userspace/libscap/scap_udig.c

```diff
@@ -147,43 +492,46 @@ int32_t udig_alloc_ring_descriptors(int* ring_descs_fd,
 {
 	uint32_t mem_size = sizeof(struct ppm_ring_buffer_info) + sizeof(struct udig_ring_buffer_status);
 
-	//
-	// First, try to open an existing ring
-	//
+#ifdef UDIG
+	// This is a producer, let's tell it to open an existing
+	// ring using ud_shm_open
 	*ring_descs_fd = ud_shm_open(UDIG_RING_DESCS_SM_FNAME, O_RDWR, 0);
 	if(*ring_descs_fd < 0)
+	{
+			snprintf(error, SCAP_LASTERR_SIZE, "udig_alloc_ring_descriptors shm_open error: %s\n", strerror(errno));
+			udig_memfd_shm_unlink(UDIG_RING_DESCS_SM_FNAME);
+			return SCAP_FAILURE;
+	}
```

Comment:
this is what happens when the consumer is not running? to keep the current behavior, ideally the producer would silently drop all events *until* a consumer becomes available

not sure how to do that cleanly without trying to connect on every event, maybe reconnect on a 500ms-or-so timer? (I'm sure 500ms is both too high for some use cases and too low for others)

Reply:
This is outdated too. When the consumer is not running this is what happens:

```
sudo  ./userspace-example
error initializing userspace data structures: udig_alloc_ring shm_open error: Connection refused
```

While this is what happens when you don't have the necessary permissions to connect to the socket that gives you the file descriptors (notice that sudo is missing here):


```
./userspace-example 
error initializing userspace data structures: udig_alloc_ring shm_open error: Permission denied
```

I think that in the first case we can do as you were suggesting (something like an exponential backoff for re-connecting). However I'm not sure that this must be handled by scap.

I would prefer the implementer of the producer to decide what to do when they meet the connection refused error and act based on that.

This was a source of confusion for me given that some producers are using the `res < 0` notation implying that they will get a negative int from those functions. However a negative int is never given since `SCAP_FAILURE` equals to 1. Example in [pdig](https://github.com/falcosecurity/pdig/blob/d98702b8f424c91dbdaf7910d86d46328f9fa596/pdig_compat.c#L51-L54)

In my example I'm able to get the "Permission denied" message because I'm checking the error buffer:

```
  char error_buf[256];
  int res;
  res = userspace_init(error_buf);
  if (res == SCAP_FAILURE) {
    printf("error initializing userspace data structures: %s", error_buf);
    return 1;
  }
```

A check to `errno` leads to the same results.

Given all of this, in my opinion scap should not be responsible for this additional duty, since it's very easy to do outside and for the producer writers it becomes easier to deal with failure rather than waiting in a loop.

Reply:
I'm not saying this is libscap's core responsibility; this patch is the only situation where a producer fails because there's no consumer so I'd say it's the responsibility of the udig producer.

This is udig-specific but since ~all the code is shared now, it probably has to live in libscap (or maybe pdig).

Reply:
That's exactly what I want to say, if it's responsibility of the udig producer, and producers are implemented outside this repository I would let the producer themselves implement that mechanism?

I would not put this in pdig since that is a specific producer implementation.